Chime/Parser.m

```diff
@@ -0,0 +1,268 @@
+
+  NSData *data = [[NSData alloc] initWithContentsOfFile:@(filepath)];
+
+  if (data == nil) {
+    NSLog(@"Could not read keywords from file \"%s\"", filepath);
+  }
+
+  NSString *stringKeywords =
+      [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+
+  NSArray *keywordsArr = [stringKeywords componentsSeparatedByString:@"\n"];
+
+  for (uint64_t i = 0; i < OP_TOTAL; i++) {
+    [_keywords setObject:@(i) forKey:[keywordsArr objectAtIndex:i]];
+  }
+}
+
+BOOL isStringStartingWith(NSString *line, NSString *target) {
+  return [line hasPrefix:target];
+}
+
+static inline BOOL isLineEmpty(id line) {
+  return line == nil ||
+         ([line respondsToSelector:@selector(length)] &&
+          [(NSData *)line length] == 0) ||
+         ([line respondsToSelector:@selector(count)] &&
+          [(NSArray *)line count] == 0);
+}
+
+BOOL isNumberLine(NSString *line) { return isStringStartingWith(line, @"#"); }
+
+BOOL isLabelLine(NSString *line) { return isStringStartingWith(line, @":"); }
+
+BOOL isVariableLine(NSString *line) {
+  return isStringStartingWith(line, @">") &&
+         isStringStartingWith(tokenAt(line, 1), @"#");
+}
+
+BOOL isAddressLine(NSString *line) { return isStringStartingWith(line, @"&"); }
+
+BOOL isDerefLine(NSString *line) { return isStringStartingWith(line, @"*"); }
+
+NSString *tokenAt(NSString *string, NSInteger index) {
+  NSArray *line = [string componentsSeparatedByString:@" "];
+  if ([line count] <= index) {
+    @throw [NSException
+        exceptionWithName:@"Line index out of bounds"
+                   reason:[NSString
+                              stringWithFormat:@"Could not find token at "
+                                               @"position %lu in line \"%@\"",
+                                               index, line]
+                 userInfo:nil];
+  } else {
+    return [line objectAtIndex:index];
+  }
+}
+
+NSString *popFirstChar(NSString *string) {
+  return [string substringFromIndex:1];
+}
+
+- (void)handleLabelLine:(NSString *)line
+        withLineCounter:(NSUInteger)nonEmptyLinesCounter {
+  NSString *labelColon = tokenAt(line, 0);
+  NSString *label = popFirstChar(labelColon);
+  if (findInEnumerator([_labels keyEnumerator], label)) {
+    @throw [NSException
+        exceptionWithName:@"Label Collision"
+                   reason:[NSString
+                              stringWithFormat:@"Found label \"%@\" more than "
+                                               @"once. Lines %lu and %@.",
+                                               label,
+                                               [[_labels objectForKey:label]
+                                                   integerValue],
+                                               @(_counter)]
+                 userInfo:nil];
+  } else {
+    [_labels setObject:@(nonEmptyLinesCounter) forKey:label];
+  }
+}
+
+// TODO: Investigate modal code sections for compile time area and run time area
```

Comment:
I kind of forgot what this comment was about.

Reply:
I vaguely remember being about having areas of chime code in which we do some things and when going to other section we do something else?

Reply:
This is lost in time like the ancients, so I will remove it